src/services/gquery.js

```diff
@@ -0,0 +1,246 @@
+        for (let subscription of this.subscriptions.slice(1)) {
+            merge(root, subscription.query);
+        }
+        this.query = root;
+    
+        // TODO - remove
+        this.print();
+    }
+
+    register(view, callbacks) {
+        /**
+         * Register a view to this instance.
+         * @param {Vue.View} view - The view to register.
+         * @param {Object} callbacks - A map of the form {name: fcn}.
+         */
+        this.views.push({
+            id: view.viewID,
+            view,
+            ...callbacks
+        });
+    }
+
+    unregister(view) {
+        /**
+         * Unregister a view (all subscriptions will be dropped.
+         * @param {Vue.View} view - The view to unregister.
+         */
+        this.views = this.views.filter(
+            v => v.view != view
+        );
+        this.subscriptions = this.subscriptions.filter(
+            s => s.view != view
+        );
+        this.recompute();
+    }
+
+    subscribe(view, query) {
+        /**
+         * Subscribe a new query.
+         * @param {Vue.View} view - The view to subscribe the query to.
+         * @param {string} query - The query to subscribe.
+         * @return {number} The subscription ID (used for un-subscribing).
+         */
+        var id = Math.random();
+        this.subscriptions.push({
+            id,
+            view,
+            query: parse(query),
+            active: false
+        });
+        this.recompute();
+        return id;
+    }
+
+    unsubscribe(id) {
+        /**
+         * Un-subscribe from a query.
+         * @param {number} id - The subscription ID to un-subscribe.
+         */
+        this.subscriptions = this.subscriptions.filter(
+            s => s.id != id
+        );
+        this.recompute();
+    }
+
+    request() {
+        /**
+         * Perform a REST GraphQL request for all subscriptions.
+         */
+        console.log('graphql request:', this.query);
+        return this.apolloClient.query({
+            query: this.query,
+            fetchPolicy: 'no-cache'
+        }).then((response) => {
+            // commit results
+            store.dispatch(
+                'workflows/set',
+                response.data.workflows
+            );
+            // set all subscriptions to active
+            this.subscriptions
+                .filter(s => s.active === false)
+                .forEach(s => s.active = true);
+            // run callback functions on the views
+            this.callbackActive();
+        }).catch((error) => {
+            store.dispatch(
+                'setAlert',
+                new Alert(error.message, null, 'error')
+            );
+        });
+    }
+
+    callbackActive() {
+        /**
+         * Run callback functions registered by views.
+         */
+        // activeCallback - called when all subscriptions are active
+        this.views
+            .filter(v => v.activeCallback)
```

Comment:
We could have a `View` (or a subtype) that has the callback when active... or maybe use one of the component lifecycle methods (beforeCreate, beforeDestroy, etc)...

Reply:
There may be a few callbacks, we will definitely need one for sharing selections (e.g. double click on a task in the graph view to get the tree view to select it too).